### PR TITLE
Keep a compiled forward alive when Dynamo gives up on it

### PR DESCRIPTION
**Pull request by danielhanchen**
Four related failures, all of them a compiled forward dying rather than degrading.

### `requires_grad` hooks cannot be traced

The gradient-enabling hooks mutate `requires_grad`, which Dynamo refuses:

```
Unsupported: Unsupported Tensor.requires_grad_() call
```

They are autograd bookkeeping rather than compute, one tensor once per forward, so they now run eagerly with a graph break around them. On a torch without `_dynamo` the decorator degrades to a no-op, so eager users and older builds are untouched.

The wrapper copies `__name__` / `__qualname__` back, because `register_other_hooks()` identifies our hooks by matching those names. Without it the hooks stop being recognised as ours and get registered on top of themselves. torch 2.9 happens to carry them through `disable()`, but that is an internal detail of a private module and this has to hold from torch 2.6 up.

### `fullgraph = True` turns cache exhaustion into a hard failure

```
FailOnRecompileLimitHit: recompile_limit reached with fullgraph=True
```

which is a performance problem presented as a training failure. The patched function now falls back to eager. Only cache exhaustion and our own disabled-hook graph break are caught; any other graph break under `fullgraph` still raises exactly as before.

The exception classes are looked up by name because torch 2.6 through 2.11 do not agree on which of `FailOnRecompileLimitHit`, `RecompileLimitExceeded` and `CacheLimitExceeded` exist, and a missing one must not stop the tuple from being built.

### Our own disabled hook breaking our own graph

The `torch.compiler.disable`d hooks above can be invoked from inside a `fullgraph = True` region, and Dynamo refuses:

```
Unsupported: Skip calling `torch.compiler.disable()`d function
```

Matched narrowly on that signature, so any other graph break still raises.

### The fallback has to latch, and this is the subtle one

Non-reentrant activation checkpointing recomputes each packed forward during backward and compares the saved intermediates. A pack compiled and then recomputed eagerly aborts the backward with "Something went unexpectedly wrong in activation checkpoint".

Per-call retry does **not** fix this, and it was tried first. The pack and the recompute run under different guards (grad mode differs, and the recompute is inside backward), so the compiler can succeed for one and raise for the other in either direction. Latching makes every call after the first failure eager, so every later pack and recompute agree.

That leaves exactly one mixed step, the one during which the latch flips. `force_eager_fallback()` closes it: unsloth calls it when a backward dies inside activation checkpointing, so the retry of that step is consistent. It returns a count of live wrappers rather than of changes, because the wrapper that caused the trouble has already latched itself and "how many changed" would be 0 in precisely the case that matters. `only_if_already_triggered` refuses to switch off compilation for a model that never fell back, so a caller that gets 0 has learned the checkpoint assertion was not a compile-mode flip and should re-raise rather than retry.

The wrapper registry holds weak references, so a patched-then-discarded model does not keep its functions alive, and the registry cannot report a switch for something nobody is calling any more.

### Tests

**560 passed, 37 skipped, 0 failed.**

- 54 new across four files: the hook decorator, the recompile-limit fallback, the disabled-hook graph break, and the latch itself (including that the latch is per wrapper and not global).
- 506 passed / 37 skipped across 15 existing suites touching compile fallback, requires_grad hooks, activation checkpointing and recompute, and `patch_function`.

Both `test_the_fallback_latches` tests carry an explicit note not to reverse the assertion, because an earlier revision of this work asserted non-latching and per-call retry is what the activation-checkpoint failure above rules out.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/temporary_patches/utils.py [hunk 1]
```diff
@@ -661,6 +663,178 @@ def _get_unique_storage_name(
...
+
+    Matched narrowly on that signature: any other graph break under fullgraph
+    must still raise, since those point at real problems.
+    """
+    try:
+        text = str(exc)
+    except Exception:
+        return False
+    return ("torch.compiler.disable" in text
+            and "Skip calling" in text)
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/utils.py]: Only patch_function(..., fullgraph = True) targets get this wrapper, which is nine leaves (Gemma3/Gemma4/Gemma3n RMSNorm and scaled word embedding, AltUp, Qwen3MoeRMSNorm, timm ConvNormAct), and grepping installed transformers and timm for compiler.disable / _dynamo.disable finds no disabled callee reachable from any of them. A wrong match would also just run eager_func, the same Python uncompiled, so there are no numerics to mask.

File: unsloth_zoo/temporary_patches/utils.py [hunk 2]
```diff
@@ -661,6 +663,178 @@ def _get_unique_storage_name(
...
+    assertion was not a compile-mode flip, and should re-raise rather than
+    retry.
+    """
+    live = [w for w in (ref() for ref in _EAGER_FALLBACK_WRAPPERS)
+            if w is not None]
+    if only_if_already_triggered and not any(
+            w._unsloth_fallback_state["eager"] for w in live):
+        return 0
+    for w in live:
+        w._unsloth_fallback_state["eager"] = True
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/temporary_patches/utils.py:910]: patch_function does setattr on the transformers/timm class at utils.py:894, so two live Gemma3 models share the identical wrapper object and there is no per-model wrapper to scope the registry to without making patching per-instance. Latching to eager also changes speed only, not numerics, and force_eager_fallback is opt-in with that global contract stated in its docstring at utils.py:811-826.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
